editor/modules/aariba/interfaces.ts

```diff
@@ -20,16 +20,22 @@ export interface TextUpdate {
     selection: any[];
 }
 
-export interface BehaviourAction {
-    (state: string, action: 'insertion', editor: AceAjax.Editor,
-     session: AceAjax.IEditSession, range: AceAjax.Range): TextUpdate;
-    (state: string, action: 'deletion', editor: AceAjax.Editor,
-     session: AceAjax.IEditSession, text: string): AceAjax.Range;
-    (state: string, action: string, editor: AceAjax.Editor,
-     session: AceAjax.IEditSession, range_or_text: AceAjax.Range | string): any;
+export interface BehaviourAction {}
+
+export interface BehaviourActionDel extends BehaviourAction {
+    (this: any, state: string, action: string, editor: AceAjax.Editor,
+     session: AceAjax.IEditSession, range: AceAjax.Range): TextUpdate | undefined;
+}
+
+export interface BehaviourActionIns extends BehaviourAction {
+    (this: any, state: string, action: string, editor: AceAjax.Editor,
+     session: AceAjax.IEditSession, text: string): AceAjax.Range | undefined;
 }
 
 export interface IsBehaviour extends IsType<IsBehaviour> {
+    // This does not typecheck anything...
```

Comment:
So I'm confused with the upgrade of TypeScript because the code no longer work as before (but even before it seems that it was broken). The idea is to restrict the values that can be passed as a third argument using the second one. I think I have seen a similar thing in the dom API type definition. However this one does not work as the captured variant is always the last one. I don't know if this is a bug from my part or from TypeScript.

I guess we can add an issue to investigate that later.